fix(#544): gridfs by mongodb GridFSBucket

vul3q04 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mongo-express/mongo-express/1811)
- - -
<!-- Reviewable:end -->
Since gridfs-stream has been deprecated. the bucket should be move to GridFSBucket from mongodb.

Solve the endpoint function: viewBucket, addFile, getFile and deleteFile

related #554

Commit:
Fix: using bucket for gridfs

vul3q04 committed:
Fix: use GridFSBucket implement upload & delete

vul3q04 committed:
Fix: remove unused lib

shakaran committed:
fix: migrate GridFS to the driver's GridFSBucket

GridFS was completely broken. Opening any bucket returned a 500:

  TypeError: mongo.connections[req.dbName].collection is not a function

router.js used connections[dbName].collection(...) where the rest of the file
correctly uses connections[dbName].db.collection(...). Nothing downstream ever
ran, which is why the rest went unnoticed.

Underneath that, lib/routes/gridfs.js was built on gridfs-stream — sunset in
2015, flagged by yarn audit, and its last release was 1.1.1 — and on
mongo.GridStore, which the MongoDB driver removed in v4. On the 6.x driver in
use here mongo.GridStore is undefined, so every file handler would have thrown
as soon as it was reached. Both are replaced by the driver's own GridFSBucket
and the dependency is dropped.

Also renames req.files to req.connFiles: express-fileupload owns req.files, and
the bucket listing was overwriting it.

Reworked from the original PR, which regressed a few things:

- res.redirect('back') no longer resolves in Express 5; it sends a literal
  'Location: back'. master had already migrated these to the Referrer form.
- getFile dropped Content-Type, Content-Disposition and the missing-file check,
  so downloads arrived unnamed, untyped, and a bad id streamed an empty 200.
- addFile dropped the uploaded file's content type.
- Uploads and deletes redirected on a 500ms setTimeout rather than awaiting the
  operation, racing on slow writes and swallowing failures.
- session.error was called as a function in the three not-implemented handlers,
  which is a string.

Verified end to end against a real MongoDB: listing a bucket now returns 200
where master returns 500, download returns the file intact with its filename,
upload stores it with its content type preserved, and delete removes it leaving
no orphaned chunks.

shakaran committed:
Merge fongting's original commits back into the branch

I force-pushed earlier and collapsed 6ce2161 ('Fix: remove unused lib') out of
the branch history. That was wrong: the work is fongting's and the attribution
belongs on the branch. Merging it back restores it without another force push;
the file contents are unchanged.

shakaran committed:
fix: redirect to the configured base URL in the GridFS bucket handler

CodeQL flagged the redirect as request-controlled data reaching res.redirect:
res.locals.baseHref is built from req.originalUrl. Same treatment as #1900 —
config.site.baseUrl is server configuration the caller cannot influence.

shakaran reviewed (approved):
Thanks @vul3q04 — this turned out to matter far more than the diff suggests, because **GridFS is completely broken on master**.

Opening any bucket returns a 500:

```
TypeError: mongo.connections[req.dbName].collection is not a function
```

`router.js` called `connections[dbName].collection(...)` where the rest of the file correctly uses `connections[dbName].db.collection(...)`. Nothing downstream ever executed, which is why the deeper problems went unnoticed: `gridfs-stream` was sunset in **2015** (last release 1.1.1, flagged by `yarn audit`), and `mongo.GridStore` was **removed from the MongoDB driver in v4** — on the 6.x driver in use it is `undefined`, so every file handler would have thrown the moment it was reached. Your migration to the driver's own `GridFSBucket` is the right fix, and it drops the dependency.

Also correct: renaming `req.files` to `req.connFiles`. `express-fileupload` owns `req.files`, and the bucket listing was overwriting it.

**Reworked a few regressions on top:**

| | |
|---|---|
| `res.redirect('back')` | Express 5 sends a literal `Location: back`; master had already migrated these to the Referrer form, and the PR reintroduced it |
| `getFile` | dropped `Content-Type`, `Content-Disposition` and the missing-file check — downloads arrived unnamed and untyped, and a bad id streamed an empty 200 |
| `addFile` | dropped the uploaded file's content type |
| `setTimeout(500)` | redirected without awaiting the write, racing on slow writes and swallowing failures |
| `req.session.error(...)` | called as a function in the three not-implemented handlers; it is a string |

**Verified end to end against a real MongoDB**, not just CI:

| Operation | master | here |
|---|---|---|
| list bucket | **500** | 200 |
| download | — | file intact, filename preserved |
| upload | — | stored with `contentType=text/plain` |
| delete | — | removed, no orphaned chunks |

Finally, CodeQL flagged the bucket handler's redirect as request-derived data reaching `res.redirect`; it now uses `config.site.baseUrl`, matching #1900. The two remaining bot comments are anchored to superseded commits.

Lint clean, 96 passing, CI green on Node 22, 24 and 26.

One caveat for the record: a nonexistent bucket returns 200 rather than redirecting, because `toArray()` on a missing collection returns `[]` instead of throwing, so that error branch is effectively unreachable. That is pre-existing on master and left alone here.

shakaran committed:
test: cover the GridFS routes

GridFS had no automated coverage, which is how it came to be entirely broken on
master without anyone noticing. Four cases over a real bucket: listing, download,
upload and delete.

Verified they actually catch the regressions rather than just passing:

- reverting the router.js '.db' fix fails all four
- removing the Content-Disposition header fails the download case, which is one
  of the things the original PR dropped

The bucket is seeded before createServer(), because the app caches its collection
list on connect and colsToGrid() builds res.locals.gridFSBuckets from it.

shakaran commented:
Added automated coverage, since GridFS having none is precisely how it came to be broken on master without anyone noticing.

`test/lib/routers/gridfsSpec.js` — four cases over a real bucket: listing, download, upload and delete.

Rather than assume they are meaningful, I checked that they fail when the bugs are reintroduced:

| Reintroduced | Result |
|---|---|
| `connections[db].collection()` without `.db` (master's bug) | **all four fail** |
| dropping `Content-Disposition` (one of the original PR's regressions) | the download case fails |

One implementation note: the bucket is seeded before `createServer()`, because the app caches its collection list on connect and `colsToGrid()` builds `res.locals.gridFSBuckets` from it.

100 passing (up from 96), lint clean, CI green on Node 22, 24 and 26. The two remaining CodeQL comments are anchored to superseded commits (`6ce2161`, `8b6d721`); the current head is `372f700`.

Approval stands.
